File: packages/dart/ethereal_genui_core/lib/src/genui_options.dart
/// A normalized choice/option. The model may emit a plain string (`"Daily"`) OR
/// an object (`{"label":"Daily","value":"daily"}`, `{"label":"x","checked":true,
/// "send":"y"}`). Renderers must never show the raw map — always read [label]
/// for display and [value] for the action.
class GenUiOption {
  const GenUiOption({required this.label, required this.value, this.checked = false});
  final String label;
  final String value;
  final bool checked;
}

/// Normalize a model-supplied list of options/items into [GenUiOption]s,
/// tolerating both string and object forms.
List<GenUiOption> genUiOptions(dynamic raw) {
  final list = raw is List ? raw : const [];
  final out = <GenUiOption>[];
  for (final e in list) {
    if (e is Map) {
      final label =
          (e['label'] ?? e['text'] ?? e['title'] ?? e['name'] ?? e['value'] ?? '')
              .toString();
      final value = (e['value'] ?? e['send'] ?? e['label'] ?? label).toString();
      out.add(GenUiOption(label: label, value: value, checked: e['checked'] == true));
    } else {
      final s = e.toString();
      out.add(GenUiOption(label: s, value: s));
    }
  }
  return out;
}

File: packages/js/ethereal-genui-core/src/genui_options.ts
/// A normalized choice/option. The model may emit a plain string ("Daily") OR
/// an object ({"label":"Daily","value":"daily"}, {"label":"x","checked":true,
/// "send":"y"}). Renderers must never show the raw map — always read label
/// for display and value for the action.

export interface GenUiOption {
  readonly label: string
  readonly value: string
  readonly checked: boolean
}

/// Normalize a model-supplied list of options/items into GenUiOptions,
/// tolerating both string and object forms.
export function genUiOptions(raw: unknown): GenUiOption[] {
  if (!Array.isArray(raw)) return []
  const out: GenUiOption[] = []
  for (const e of raw) {
    if (e !== null && typeof e === 'object' && !Array.isArray(e)) {
      const map = e as Record<string, unknown>
      const label = String(
        map['label'] ?? map['text'] ?? map['title'] ?? map['name'] ?? map['value'] ?? ''
      )
      const value = String(map['value'] ?? map['send'] ?? map['label'] ?? label)
      out.push({ label, value, checked: map['checked'] === true })
    } else {
      const s = String(e)
      out.push({ label: s, value: s, checked: false })
    }
  }
  return out
}
